### PR TITLE
E2E test results and concurrency

### PR DESCRIPTION
**Pull request by jwair**
- Introduced a test splitter that finds all tests in a package (e2e in our case) and creates N files each with a regex that only runs 1/N of the tests.
- Added parallelism to the e2e-tests step in CircleCI. 3 jobs run in parallel, each running a different set of tests.
- Switched from go test to gotestsum for running e2e tests, unit and integration tests
   - Cleaner colorful output
   - Test results xml generation
- Added a step to upload test-results to CircleCI
- Refactored circleci config
- Removed obsolete go modules cache

### CODE REVIEW

File: build-scripts/split-tests.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    local output_dir=$3
+
+    local total_lines
+    local lines_per_file
+
+    total_lines=$(wc -l < "${input_file}" | xargs)
+    ((lines_per_file = (total_lines + chunks_count - 1) / chunks_count))
+
+    echo ">> Tests total count: ${total_lines}"
+    mkdir -p "${chunks_dir}"
```
Comment (hashi-derek) [inline build-scripts/split-tests.sh:26]: Is `chunks_dir` supposed to be a reference to the local `output_dir` variable?

Comment (jwair) [reply build-scripts/split-tests.sh]: Addressed in https://github.com/hashicorp/consul-terraform-sync/pull/869/commits/cc62f5479ab0303000a51d0e40c7a9a8fb3790bb

File: Makefile [hunk 2]
```diff
@@ -55,9 +55,16 @@ test-setup-e2e: dev
 # test-e2e-cirecleci does e2e test setup and then runs the e2e tests
 test-e2e-cirecleci: test-setup-e2e
 	@echo "==> Testing ${NAME} (e2e)"
-	@go test ./e2e -race -count=1 -timeout=900s -tags=e2e ./... ${TESTARGS}
+	@echo "Tests regex: $(shell cat ".build/chunks/chunk.${CHUNK_INDEX}")"
+	@gotestsum --format testname --junitfile .build/test-results/e2e-tests-${CHUNK_INDEX}.xml -- \
+		./e2e -race -count=1 -timeout=600s -tags=e2e -run="$(shell cat ".build/chunks/chunk.${CHUNK_INDEX}")" ${TESTARGS}
 .PHONY: test-e2e-cirecleci
 
+test-unit-cirecleci:
```
Comment (jwair) [inline Makefile:63]: TODO: Get rid of this and make the gotestsum in integration test make target

File: .circleci/config.yml [hunk 3]
```diff
@@ -41,48 +75,58 @@ jobs:
...
+      - setup-vault
       - run: |
           go test -count=1 -timeout=80s -tags 'integration vault' ./... -run Vault
-      - save_cache:
-          key: ct-modcache-v2-{{ checksum "go.mod" }}
-          paths:
-            - /go/pkg/mod
 
-  e2e_tests:
+  unit-tests:
```
Comment (jwair) [inline .circleci/config.yml:88]: TODO: Remove

File: .circleci/config.yml [hunk 4]
```diff
@@ -215,28 +231,28 @@ jobs:
...
-  version: 2
   build-test:
     jobs:
       - go-checks
       - terraform-checks
-      - compile_weekly_tests
-      - unit_integration_tests
-      - e2e_tests
+      - compile-weekly-tests
+      - unit-tests
```
Comment (jwair) [inline .circleci/config.yml:243]: TODO: Remove

File: Makefile (hunk 2 above)
Comment (jwair) [reply Makefile]: Addressed in https://github.com/hashicorp/consul-terraform-sync/pull/869/commits/b6e8e0ee61a8137207278705c7c473ebb52d612e

File: .circleci/config.yml (hunk 3 above)
Comment (jwair) [reply .circleci/config.yml]: Addressed in https://github.com/hashicorp/consul-terraform-sync/pull/869/commits/b6e8e0ee61a8137207278705c7c473ebb52d612e

File: .circleci/config.yml (hunk 4 above)
Comment (jwair) [reply .circleci/config.yml]: Addressed in https://github.com/hashicorp/consul-terraform-sync/pull/869/commits/b6e8e0ee61a8137207278705c7c473ebb52d612e

**Review by mkam — APPROVED** [review]
Looks good to me, just left one comment about the build directory! Thanks for all the effort you put in to improve our CI testing experience!

File: Makefile [hunk 5]
```diff
@@ -35,28 +35,32 @@ dev:
...
-# test-integration runs the test suite and integration tests
-test-integration:
-	@echo "==> Testing ${NAME} (test suite & integration)"
-	@go test -count=1 -timeout=80s -tags=integration -cover ./... ${TESTARGS}
-.PHONY: test-all
+# test-unit-and-integration runs the unit and integration tests
+test-unit-and-integration:
+	@echo "==> Testing ${NAME} (unit & integration tests)"
+	@mkdir -p .build/test-results
+	@gotestsum --format testname --junitfile .build/test-results/unit-and-integration-tests.xml -- \
```
[the remark this answers is not in the document]
Comment (jwair) [reply Makefile:48]: Addressed in https://github.com/hashicorp/consul-terraform-sync/pull/869/commits/df0cabe603c2ae7fb894a68fab2f224289109e4c

File: .circleci/config.yml [hunk 6]
```diff
@@ -8,29 +8,64 @@ executors:
     docker:
       - image: docker.mirror.hashicorp.services/cimg/go:${GO_VERSION}
     environment:
-      CONSUL_VERSION: "1.10.3"
       GOMAXPROCS: 4
       GO111MODULE: "on"
       GOPROXY: https://proxy.golang.org/
       GO_VERSION: 1.16
+      CONSUL_VERSION: "1.10.3"
```
[the remark this answers is not in the document]
Comment (jwair) [reply .circleci/config.yml:15]: Ticket created for this item

**Pull request merged**